### ISSUE TITLE
Test data examples potentially confusing?

### ISSUE THREAD

**Issue by tretherington** [r-spatialecology/shar#79]
Regarding @openjournals/joss-reviews#3811 I'm wondering if you could improve the test data examples slightly to avoid any confusion.

Ideally I think I should be able to run test data and get exactly the same results to confirm everything is working properly.  This is a bit challenging for you as the software uses data randomisation so you are unlikely to get **exactly** the same results.  However, I think it would be helpful if the conclusions of the results stays the same.

Everything seems to be fine with 

``` r
results_habitat_association(pattern = species_a, raster = torus_trans)
## > Input: randomized raster | Quantile thresholds: negative < 0.025 - positive > 0.975
##   habitat count lo hi significance
## 1       1     9  2 14         n.s.
## 2       2    25  9 24     positive
## 3       3    27 11 26     positive
## 4       4     0 11 29     negative
## 5       5    12  5 18         n.s.
```

for which I get identical output in R.

But for 

```r
results_habitat_association(pattern = reconstruction, raster = landscape_classified)
## > Input: randomized point pattern | Quantile thresholds: negative < 0.025 - positive > 0.975
##   habitat count    lo    hi significance
## 1       1     8  1.00 23.35         n.s.
## 2       2    22 25.85 49.10     negative
## 3       3    33 40.80 68.05     negative
## 4       4    19 44.00 71.10     negative
## 5       5   118 20.85 60.05     positive
```

I get the following result 

```r
results_habitat_association(pattern = reconstruction, raster = landscape_classified)
> Input: randomized point pattern | Quantile thresholds: negative < 0.025 - positive > 0.975
  habitat count    lo    hi significance
1       1     8 27.90 48.10     negative
2       2    22 52.90 77.05     negative
3       3    33 29.00 57.30         n.s.
4       4    19 24.90 46.10     negative
5       5   118 11.85 26.00     positive
```

that has quite differrent lo and hi values, and different significance results for habitat classes 1 and 3.

I am assuming that all is running correctly, but I feel a bit nervous that my conclusion is different to that of the test data example.

I would suggest you:

1. Double-check that the output in the README is correct and up to date - I suppose there is a chance you changes the test data but forgot to update the README output.
2. Consider setting the random seed `set.seed()` as part of the example code so that the results will be the same.
3. Perhaps manipulate the test data such that the while the exact numbers may vary, at least the significance of the results is consistent.

**Comment by tretherington** [comment]
Regarding [https://github.com/openjournals/joss-reviews/issues/3811](https://github.com/openjournals/joss-reviews/issues/3811) I still think the test data examples are confusing.

The README has:

``` r
results_habitat_association(pattern = species_a, raster = torus_trans)
## > Input: randomized raster
## > Quantile thresholds: negative < 0.025 || positive > 0.975
##   habitat count lo hi significance
## 1       1    35 11 32     positive
## 2       2    44 22 50         n.s.
## 3       3    36 18 45         n.s.
## 4       4     4 24 53     negative
## 5       5    73 51 85         n.s.
```

But when I run this example I get:

``` r
> results_habitat_association(pattern = species_a, raster = torus_trans)
> Input: randomized raster | Quantile thresholds: negative < 0.025 - positive > 0.975
  habitat count lo hi significance
1       1     9  2 14         n.s.
2       2    25  9 24     positive
3       3    27 11 26     positive
4       4     0 11 29     negative
5       5    12  5 18         n.s.`
```
So straight away I can see the counts are different, but these should be identical as these just reflect the counts of the `species_a` in the `landscape` both of which are fixed.  So the test data is not consistent with the README example, and I find this quite confusing as I am now worried that something is wrong either with the package or with what I am doing.

The significance results are also confusing to me as while I do have the significant negative result for habitat 4 as the README describes, I also have significant positive relationships for habitat 2 and 3.  We had discussed previously that this could be a result of randomness in the processing, but now I understand the torus translation process, this is a deterministic and systematic process so the results should be identical every time - and indeed they are for me.

Rather than continuing to try and manipulate the test data, I would suggest changing the example to increase the significance level so that the only significant relationship is the negative one for habitat 4 that you describe in the README:

```r
> results_habitat_association(pattern = species_a, raster = torus_trans, significance_level = 0.01)
> Input: randomized raster | Quantile thresholds: negative < 0.005 - positive > 0.995
  habitat count   lo hi significance
1       1     9 1.00 15         n.s.
2       2    25 7.00 26         n.s.
3       3    27 8.98 29         n.s.
4       4     0 7.00 31     negative
5       5    12 4.00 20         n.s.
```

The same is true of the other example where in the README you specify that `species_b` should have a positive association to habitat 5 and show this:

```r
results_habitat_association(pattern = reconstruction, raster = landscape_classified)
## > Input: randomized pattern
## > Quantile thresholds: negative < 0.025 || positive > 0.975
##   habitat count   lo    hi significance
## 1       1     4  2.9 17.55         n.s.
## 2       2    10 31.0 51.55     negative
## 3       3    13 18.0 31.55     negative
## 4       4    15 22.0 36.00     negative
## 5       5    98 26.0 47.00     positive
```

But when I run the example I get this:

```r
> results_habitat_association(pattern = reconstruction, raster = landscape_classified)
> Input: randomized point pattern | Quantile thresholds: negative < 0.025 - positive > 0.975
  habitat count lo    hi significance
1       1     8 11 19.55     negative
2       2    22 66 76.00     negative
3       3    33 36 47.55     negative
4       4    19 35 49.55     negative
5       5   118 25 37.00     positive
```

So again the counts do not match, so again, I have to query if I'm using the correct test data.  And while I do have the significant positive association to habitat 5, I also have significant negative associations to all other habitats.  Now this is a stochastic process, so I wouldn't expect to get identical results (and thanks for adding that note to the README 👍) but I would hope to get the same *result*, by which I mean my conclusions from the analysis should be consistent even if the exact numbers fluctuate a bit.

As a pragmatic solution, I would suggest that rather than tweaking the test data for `species-b` you could just modify the text in the README to say that there is a significant positive association with habitat 5, and significant negative associations with all other habitats.  Then it would just be a case of ensuring that the results match, and from my experience of stochastic modelling this can be achieved through a combination of either increasing the significance threshold, or by increasing the number of simulations to the point where the result stabilises.  It looks like you are actually fine, as a second run does produce slightly different lo and hi values, but the significant associations are the same, so I think you are good if you just update the count column, and modify the preamble text to note the negative associations with habitats 1 to 4. 

```r
> Input: randomized point pattern | Quantile thresholds: negative < 0.025 - positive > 0.975
  habitat count lo    hi significance
1       1     8 11 19.55     negative
2       2    22 66 76.00     negative
3       3    33 36 47.55     negative
4       4    19 35 49.55     negative
5       5   118 25 37.00     positive
```

I would really encourage you to try and get these test data examples working well, as for me, my trust in a package is highly dependent on replicating test examples, and I am sure this will be true of other potential users as well.

**Comment by mhesselbarth** [comment]
> [...] So straight away I can see the counts are different, but these should be identical as these just reflect the counts of the species_a in the landscape both of which are fixed. [...]

Yes, the counts should be identical. I am not sure why they are not, I thought I fixed that. I will look into this, sorry.

> [...] The significance results are also confusing to me as while I do have the significant negative result for habitat 4 as the README describes, I also have significant positive relationships for habitat 2 and 3 [...]

This is partly due to the randomness of the methods, but not completely. It is also a "feature" of the methods. At one point, a negative association to one habitat has to lead to a positive association to another one (and vice versa). This is also  known and stated in the literature e.g. in Yamada et al. (2006): _"[...] If a species has positive association with one habitat, it necessarily has negative association with some others, and it is only a matter of the power of the test when it will become significant. [...]"_

I will add this explanation to the README and the help page of the data as well. Furthermore, I will look into the significance levels as suggested by you to ensure results are more "stable".

**References**
Yamada, T., Tomita, A., Itoh, A., Yamakura, T., Ohkubo, T., Kanzaki, M., Tan, S., Ashton, P.S., 2006. Habitat associations of Sterculiaceae trees in a Bornean rain forest plot. Journal of Vegetation Science 17, 559–566. [https://doi.org/10.1111/j.1654-1103.2006.tb02479.x](https://doi.org/10.1111/j.1654-1103.2006.tb02479.x)

**Comment by mhesselbarth** [comment]
README and paper should be updated with the correct example data now!

**Comment by tretherington** [comment]
This all looks great, many thanks. 😄

However, it looks like you've renamed or removed the `reconstruct_pattern_cluster` function that is still used in the README examples.  I used the `reconstruct_pattern` function instead which works, but you should update the examples to refer to the new function name.

**Comment by mhesselbarth** [comment]
Glad to hear things look better now. 

Yes, `reconstruct_pattern` is a new wrapper I introduced. Should be fixed in 155c008.
